compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSuspendOnInvalidTypeChecker.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirElement
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirFile
+import org.jetbrains.kotlin.fir.types.FirFunctionTypeRef
+import org.jetbrains.kotlin.fir.types.FirResolvedTypeRef
+import org.jetbrains.kotlin.fir.types.FirTypeRefWithNullability
+import org.jetbrains.kotlin.fir.visitors.FirDefaultVisitorVoid
+import org.jetbrains.kotlin.lexer.KtTokens
+
+object FirSuspendOnInvalidTypeChecker : FirFileChecker() {
+    override fun check(declaration: FirFile, context: CheckerContext, reporter: DiagnosticReporter) {
+        // Traverse the entire file to look for source type refs (i.e., Fir(Dynamic|User|Function)TypeRef).
+        declaration.accept(object : FirDefaultVisitorVoid() {
+            override fun visitElement(element: FirElement) {
+                val source = element.source ?: return
+                if (source.kind is FirFakeSourceElementKind) return
+                element.acceptChildren(this)
+            }
+
+            override fun visitResolvedTypeRef(resolvedTypeRef: FirResolvedTypeRef) {
+                // We will only get to the source type refs by visiting delegatedTypeRef.
```

Comment:
> // We will only get to the source type refs by visiting delegatedTypeRef.

That is the next line:

```kt
resolvedTypeRef.delegatedTypeRef?.accept(this)
```

Then, why do we need:
```kt
super.visitResolvedTypeRef(resolvedTypeRef)
```
?

Reply:
I had it there so the other children are visited, in this case `FirResolvedTypeRef.annotations`. Eventually execution will make its way back to `visitElement()` and those children will be visited there.

That said annotation calls can only have compile-time constants so no harm in not checking those calls, but FE1.0 still reports `WRONG_MODIFIER_TARGET` on `suspend` in `@Foo({ val x: suspend String? }())`

Reply:
I added a test for this too.